File: scripts/project-metadata/migrate-to-pidex-folder.py
#!/usr/bin/env python3
"""Migrate project-local PIDEX metadata out of wiki/ and into pidex/."""
from __future__ import annotations

import argparse
import filecmp
import json
import shutil
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

TARGET_DIRS = ["pidex/rules", "pidex/state", "pidex/config", "pidex/prompts"]


def rel(project: Path, path: Path) -> str:
    return path.relative_to(project).as_posix()


def unique_path(path: Path) -> Path:
    if not path.exists():
        return path
    stem = path.name
    parent = path.parent
    for i in range(1, 1000):
        candidate = parent / f"{stem}.{i}"
        if not candidate.exists():
            return candidate
    raise RuntimeError(f"could not find unique archive path for {path}")


def same_file_bytes(a: Path, b: Path) -> bool:
    return a.is_file() and b.is_file() and filecmp.cmp(a, b, shallow=False)


def copy_dir_no_overwrite(src: Path, dst: Path, errors: list[str], planned: list[dict[str, str]], dry_run: bool, project: Path) -> None:
    for item in sorted(src.rglob("*")):
        if item.is_dir():
            continue
        rel_item = item.relative_to(src)
        target = dst / rel_item
        planned.append({"action": "copy", "from": rel(project, item), "to": rel(project, target)})
        if target.exists() and not same_file_bytes(item, target):
            errors.append(f"conflict: {rel(project, target)} exists with different content")
            continue
        if not dry_run:
            target.parent.mkdir(parents=True, exist_ok=True)
            if not target.exists():
                shutil.copy2(item, target)


def migrate_project(project_arg: str, dry_run: bool) -> dict[str, Any]:
    project = Path(project_arg).expanduser().resolve()
    stamp = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    archive_root = project / ".wiki-migration" / "archive" / stamp / "pidex-metadata"
    result: dict[str, Any] = {
        "ok": True,
        "project": str(project),
        "dry_run": dry_run,
        "created_dirs": [],
        "migrations": [],
        "warnings": [],
        "errors": [],
    }
    if not project.exists():
        result["ok"] = False
        result["errors"].append("project root does not exist")
        return result

    for d in TARGET_DIRS:
        path = project / d
        if not path.exists():
            result["created_dirs"].append(d)
            if not dry_run:
                path.mkdir(parents=True, exist_ok=True)

    legacy_rules = project / "wiki" / "rules"
    target_rules = project / "pidex" / "rules"
    if legacy_rules.exists():
        if not legacy_rules.is_dir():
            result["errors"].append("wiki/rules exists but is not a directory")
        else:
            copy_dir_no_overwrite(legacy_rules, target_rules, result["errors"], result["migrations"], dry_run, project)
            archive_rules = unique_path(archive_root / "wiki-rules")
            result["migrations"].append({"action": "archive", "from": "wiki/rules", "to": rel(project, archive_rules)})
            if not dry_run and not result["errors"]:
                archive_rules.parent.mkdir(parents=True, exist_ok=True)
                shutil.move(str(legacy_rules), str(archive_rules))
    else:
        result["warnings"].append("no legacy wiki/rules directory")

    legacy_state = project / "wiki" / ".hygiene-state.json"
    target_state = project / "pidex" / "state" / "wiki-hygiene.json"
    if legacy_state.exists():
        if not legacy_state.is_file():
            result["errors"].append("wiki/.hygiene-state.json exists but is not a file")
        elif target_state.exists() and not same_file_bytes(legacy_state, target_state):
            result["errors"].append("conflict: pidex/state/wiki-hygiene.json exists with different content")
        else:
            result["migrations"].append({"action": "copy", "from": "wiki/.hygiene-state.json", "to": "pidex/state/wiki-hygiene.json"})
            archive_state = unique_path(archive_root / "hygiene-state.json")
            result["migrations"].append({"action": "archive", "from": "wiki/.hygiene-state.json", "to": rel(project, archive_state)})
            if not dry_run:
                target_state.parent.mkdir(parents=True, exist_ok=True)
                if not target_state.exists():
                    shutil.copy2(legacy_state, target_state)
                archive_state.parent.mkdir(parents=True, exist_ok=True)
                shutil.move(str(legacy_state), str(archive_state))
    else:
        result["warnings"].append("no legacy wiki/.hygiene-state.json")

    if result["errors"]:
        result["ok"] = False
    return result


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--json", action="store_true", help="Emit JSON only (default also emits JSON).")
    parser.add_argument("projects", nargs="+")
    args = parser.parse_args()
    results = [migrate_project(p, args.dry_run) for p in args.projects]
    payload = {"ok": all(r["ok"] for r in results), "dry_run": args.dry_run, "projects": results}
    print(json.dumps(payload, indent=2))
    return 0 if payload["ok"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: modules/pidex/project-context/scripts/project-metadata/migrate-to-pidex-folder.mjs
#!/usr/bin/env node
// Migrate project-local PIDEX metadata out of wiki/ and into pidex/.
import { existsSync, mkdirSync, readdirSync, readFileSync, renameSync, statSync, copyFileSync } from 'node:fs';
import path from 'node:path';
import process from 'node:process';

const TARGET_DIRS = ['pidex/rules', 'pidex/state', 'pidex/config', 'pidex/prompts'];
function rel(project, file) { return path.relative(project, file).replaceAll(path.sep, '/'); }
function uniquePath(file) { if (!existsSync(file)) return file; const parent = path.dirname(file); const stem = path.basename(file); for (let i = 1; i < 1000; i++) { const c = path.join(parent, `${stem}.${i}`); if (!existsSync(c)) return c; } throw new Error(`could not find unique archive path for ${file}`); }
function sameFileBytes(a, b) { try { return statSync(a).isFile() && statSync(b).isFile() && readFileSync(a).equals(readFileSync(b)); } catch { return false; } }
function walkFiles(dir) { const out = []; for (const entry of readdirSync(dir, { withFileTypes: true })) { const p = path.join(dir, entry.name); if (entry.isDirectory()) out.push(...walkFiles(p)); else if (entry.isFile()) out.push(p); } return out.sort(); }
function copyDirNoOverwrite(src, dst, errors, planned, dryRun, project) { for (const item of walkFiles(src)) { const relItem = path.relative(src, item); const target = path.join(dst, relItem); planned.push({ action: 'copy', from: rel(project, item), to: rel(project, target) }); if (existsSync(target) && !sameFileBytes(item, target)) { errors.push(`conflict: ${rel(project, target)} exists with different content`); continue; } if (!dryRun) { mkdirSync(path.dirname(target), { recursive: true }); if (!existsSync(target)) copyFileSync(item, target); } } }
function expandHome(value) { return String(value).replace(/^~(?=$|[\\/])/, process.env.HOME || ''); }
function migrateProject(projectArg, dryRun) {
  const project = path.resolve(expandHome(projectArg));
  const stamp = new Date().toISOString().slice(0, 10);
  const archiveRoot = path.join(project, '.wiki-migration', 'archive', stamp, 'pidex-metadata');
  const result = { ok: true, project, dry_run: dryRun, created_dirs: [], migrations: [], warnings: [], errors: [] };
  if (!existsSync(project)) { result.ok = false; result.errors.push('project root does not exist'); return result; }
  for (const d of TARGET_DIRS) { const p = path.join(project, d); if (!existsSync(p)) { result.created_dirs.push(d); if (!dryRun) mkdirSync(p, { recursive: true }); } }
  const legacyRules = path.join(project, 'wiki', 'rules'); const targetRules = path.join(project, 'pidex', 'rules');
  if (existsSync(legacyRules)) { if (!statSync(legacyRules).isDirectory()) result.errors.push('wiki/rules exists but is not a directory'); else { copyDirNoOverwrite(legacyRules, targetRules, result.errors, result.migrations, dryRun, project); const archiveRules = uniquePath(path.join(archiveRoot, 'wiki-rules')); result.migrations.push({ action: 'archive', from: 'wiki/rules', to: rel(project, archiveRules) }); if (!dryRun && !result.errors.length) { mkdirSync(path.dirname(archiveRules), { recursive: true }); renameSync(legacyRules, archiveRules); } } }
  else result.warnings.push('no legacy wiki/rules directory');
  const legacyState = path.join(project, 'wiki', '.hygiene-state.json'); const targetState = path.join(project, 'pidex', 'state', 'wiki-hygiene.json');
  if (existsSync(legacyState)) { if (!statSync(legacyState).isFile()) result.errors.push('wiki/.hygiene-state.json exists but is not a file'); else if (existsSync(targetState) && !sameFileBytes(legacyState, targetState)) result.errors.push('conflict: pidex/state/wiki-hygiene.json exists with different content'); else { result.migrations.push({ action: 'copy', from: 'wiki/.hygiene-state.json', to: 'pidex/state/wiki-hygiene.json' }); const archiveState = uniquePath(path.join(archiveRoot, 'hygiene-state.json')); result.migrations.push({ action: 'archive', from: 'wiki/.hygiene-state.json', to: rel(project, archiveState) }); if (!dryRun) { mkdirSync(path.dirname(targetState), { recursive: true }); if (!existsSync(targetState)) copyFileSync(legacyState, targetState); mkdirSync(path.dirname(archiveState), { recursive: true }); renameSync(legacyState, archiveState); } } }
  else result.warnings.push('no legacy wiki/.hygiene-state.json');
  if (result.errors.length) result.ok = false;
  return result;
}
function parse(argv) { const args = { dryRun: false, projects: [] }; for (const a of argv) { if (a === '--dry-run') args.dryRun = true; else if (a === '--json') {} else args.projects.push(a); } if (!args.projects.length) { console.error('usage: migrate-to-pidex-folder.mjs [--dry-run] [--json] <project>...'); process.exit(2); } return args; }
const args = parse(process.argv.slice(2));
const results = args.projects.map((p) => migrateProject(p, args.dryRun));
const payload = { ok: results.every((r) => r.ok), dry_run: args.dryRun, projects: results };
console.log(JSON.stringify(payload, null, 2));
process.exitCode = payload.ok ? 0 : 1;
